src/tools/quemap/lightmap.c

```diff
@@ -230,6 +230,21 @@ static void CalcSamplePoints(const light_info_t *l, vec_t sofs, vec_t tofs, vec_
 	}
 }
 
+/**
+ * @brief Converts world space direction into tangent space for given TBN
+ */
+static void WorldSpaceToTangentSpace(const vec3_t tangent, const vec3_t bitangent, const vec3_t normal, const vec3_t direction, vec3_t result) {
+	// temporary storage in case if direction and result are the same pointer
+	vec3_t temporary_result;
+
+	// transform it into tangent space
+	temporary_result[0] = DotProduct(direction, tangent);
+	temporary_result[1] = DotProduct(direction, bitangent);
+	temporary_result[2] = DotProduct(direction, normal);
+
+	VectorCopy(temporary_result, result);
+}
```

Comment:
This makes sense. I dig it. To bounce indirect light, you have to transform the other way, tho, right?